File: fraiseql-typescript/tests/generate-parity-schema.ts
/**
 * Generate parity schema for cross-SDK comparison.
 *
 * Usage:
 *   bun run tests/generate-parity-schema.ts
 */

import { SchemaRegistry, registerTypeFields, registerQuery, registerMutation } from "../src/index";

SchemaRegistry.clear();

// --- Types ---

registerTypeFields("User", [
  { name: "id",    type: "ID",     nullable: false },
  { name: "email", type: "String", nullable: false },
  { name: "name",  type: "String", nullable: false },
], undefined, { sqlSource: "v_user" });

registerTypeFields("Order", [
  { name: "id",    type: "ID",    nullable: false },
  { name: "total", type: "Float", nullable: false },
], undefined, { sqlSource: "v_order" });

registerTypeFields("UserNotFound", [
  { name: "message", type: "String", nullable: false },
  { name: "code",    type: "String", nullable: false },
], undefined, { isError: true, sqlSource: "v_user_not_found" });

// --- Queries ---

registerQuery("users", "User", true, false, [], undefined, {
  sqlSource: "v_user",
});

registerQuery("tenantOrders", "Order", true, false, [], undefined, {
  sqlSource: "v_order",
  inject: { tenant_id: "jwt:tenant_id" },
  cacheTtlSeconds: 300,
  requiresRole: "admin",
});

// --- Mutations ---

registerMutation("createUser", "User", false, false, [
  { name: "email", type: "String", nullable: false },
  { name: "name",  type: "String", nullable: false },
], undefined, {
  sqlSource: "fn_create_user",
  operation: "insert",
});

registerMutation("placeOrder", "Order", false, false, [], undefined, {
  sqlSource: "fn_place_order",
  operation: "insert",
  inject: { user_id: "jwt:sub" },
  invalidatesViews: ["v_order_summary"],
  invalidatesFactTables: ["tf_sales"],
});

// Output schema as JSON
const schema = SchemaRegistry.getSchema();
const output = {
  types: schema.types,
  queries: schema.queries,
  mutations: schema.mutations,
};
process.stdout.write(JSON.stringify(output, null, 2) + "\n");

File: fraiseql-python/tests/generate_parity_schema.py
#!/usr/bin/env python3
"""Generate parity schema for cross-SDK comparison.

Usage:
    uv run python tests/generate_parity_schema.py
"""

import json
import sys

sys.path.insert(0, "src")

from fraiseql.registry import SchemaRegistry  # noqa: E402
import fraiseql  # noqa: E402
from fraiseql.scalars import ID  # noqa: E402

# Reset registry to avoid contamination from previous imports
SchemaRegistry.clear()


# --- Types ---


@fraiseql.type
class User:
    id: ID
    email: str
    name: str


@fraiseql.type
class Order:
    id: ID
    total: float


@fraiseql.error
class UserNotFound:
    message: str
    code: str


# --- Queries ---


@fraiseql.query(sql_source="v_user", auto_params=True)
def users() -> list[User]:
    """List all users."""


@fraiseql.query(
    sql_source="v_order",
    inject={"tenant_id": "jwt:tenant_id"},
    cache_ttl_seconds=300,
    requires_role="admin",
)
def tenantOrders() -> list[Order]:
    """List tenant orders (admin only)."""


# --- Mutations ---


@fraiseql.mutation(sql_source="fn_create_user", operation="insert")
def createUser(email: str, name: str) -> User:
    """Create a new user."""


@fraiseql.mutation(
    sql_source="fn_place_order",
    operation="insert",
    inject={"user_id": "jwt:sub"},
    invalidates_views=["v_order_summary"],
    invalidates_fact_tables=["tf_sales"],
)
def placeOrder() -> Order:
    """Place a new order."""


# Output schema as JSON
schema = SchemaRegistry.get_schema()
output = {
    "types": schema["types"],
    "queries": schema["queries"],
    "mutations": schema["mutations"],
}
print(json.dumps(output, indent=2))
